File: src/client/js/agentguard/checkers/registry.js
"use strict";

const CHECKERS = new Map();
const DESCRIPTIONS = new Map();

let DISCOVERED = false;

function register(name, description) {
  if (!name) {
    throw new Error("checker registration name must not be empty");
  }
  return (CheckerClass) => {
    CheckerClass.prototype.name = name;
    CheckerClass.prototype.description = description;
    CHECKERS.set(name, CheckerClass);
    DESCRIPTIONS.set(name, description);
    return CheckerClass;
  };
}

function getCheckerClass(name) {
  discoverCheckers();
  return CHECKERS.get(name) || null;
}

function checkerDescriptions() {
  discoverCheckers();
  return Object.fromEntries(DESCRIPTIONS.entries());
}

function registeredCheckers() {
  discoverCheckers();
  return Object.fromEntries(CHECKERS.entries());
}

function discoverCheckers() {
  if (DISCOVERED) {
    return;
  }
  DISCOVERED = true;
  require("./llm_before/llm_input");
  require("./llm_after/llm_output");
  require("./llm_after/llm_thought");
  require("./llm_after/final_response");
  require("./tool_before/tool_invoke");
  require("./tool_after/tool_result");
}

module.exports = {
  register,
  getCheckerClass,
  checkerDescriptions,
  registeredCheckers,
  discoverCheckers,
};

File: src/client/python/agentguard/plugins/registry.py
"""Checker class registry and registration decorator."""
from __future__ import annotations

import importlib
import pkgutil
from typing import Callable

from agentguard.plugins.base import BaseChecker

_CHECKERS: dict[str, type[BaseChecker]] = {}
_DESCRIPTIONS: dict[str, str] = {}
_DISCOVERED = False


def register(name: str, description: str) -> Callable[[type[BaseChecker]], type[BaseChecker]]:
    """Register a checker class under a config-friendly name."""
    if not name:
        raise ValueError("checker registration name must not be empty")

    def _decorator(cls: type[BaseChecker]) -> type[BaseChecker]:
        if not isinstance(cls, type) or not issubclass(cls, BaseChecker):
            raise TypeError("@register can only decorate BaseChecker subclasses")
        existing = _CHECKERS.get(name)
        if (
            existing is not None
            and existing is not cls
            and existing.__module__ != cls.__module__
        ):
            raise ValueError(f"checker name already registered: {name}")
        cls.name = name
        cls.description = description
        _CHECKERS[name] = cls
        _DESCRIPTIONS[name] = description
        return cls

    return _decorator


def get_checker_class(name: str) -> type[BaseChecker] | None:
    discover_checkers()
    return _CHECKERS.get(name)


def checker_descriptions() -> dict[str, str]:
    discover_checkers()
    return dict(_DESCRIPTIONS)


def registered_checkers() -> dict[str, type[BaseChecker]]:
    discover_checkers()
    return dict(_CHECKERS)


def discover_checkers(package_name: str = "agentguard.plugins") -> None:
    """Import checker modules so @register decorators run."""
    global _DISCOVERED
    if _DISCOVERED:
        return
    _DISCOVERED = True
    package = importlib.import_module(package_name)
    package_path = getattr(package, "__path__", None)
    if package_path is None:
        return
    for module in pkgutil.walk_packages(package_path, package.__name__ + "."):
        if _should_skip(module.name):
            continue
        importlib.import_module(module.name)


def _should_skip(module_name: str) -> bool:
    leaf = module_name.rsplit(".", 1)[-1]
    return leaf in {"base", "manager", "registry"}
